src-srv/routes.ts

```diff
@@ -0,0 +1,227 @@
+import type { Repository } from 'utils/Repository.ts'
+
+/* Route types */
+interface Route {
+  path: string
+  params: string[]
+  handlers?: RouteHandlers
+}
+
+interface RouteInitContext {
+  repository: Repository
+  [key: string]: unknown
+}
+
+interface RouteContext extends RouteInitContext {
+  res: Response
+}
+
+interface RouteContentResponse {
+  payload: unknown
+  statusCode?: number
+}
+
+interface RouteStatusResponse {
+  statusCode: number
+  statusMessage: string
+}
+
+export type RouteHandler = (req: Request, context: RouteContext) => Promise<RouteContentResponse | RouteStatusResponse>
+
+export interface ApiResponseInterface {
+  isStatus: (value: unknown) => value is RouteStatusResponse
+  isContent: (value: unknown) => value is RouteContentResponse
+}
+
+interface RouteHandlers {
+  GET?: RouteHandler
+  POST?: RouteHandler
+  PUT?: RouteHandler
+  PATCH?: RouteHandler
+  DELETE?: RouteHandler
+  WEB_SOCKET?: WebsocketRequestHandler
+}
+
+type RouteMap = Record<string, Route>
+
+export const ApiResponse: ApiResponseInterface = {
+  isStatus: (value): value is RouteStatusResponse => {
+    return typeof value === 'object' && ('statusCode' in value && 'statusMessage' in value)
+  },
+  isContent: (value): value is RouteContentResponse => {
+    return typeof value === 'object' && 'payload' in value
+  }
+}
+
+/*
+ * Map directory structure under api/ to express route paths.
+ * Direcory or file names enclosed in brackets, i.e api/planning[id]/[action].ts
+ * would become "api/planning/:id/:action".
+ */
+export async function mapRoutes(apiDir: string): Promise<RouteMap> {
+  const routes: RouteMap = {}
+
+  buildRoutes(routes, apiDir)
+
+  for (const route in routes) {
+    routes[route].handlers = await importRouteHandler(routes[route].path)
+  }
+
+  return routes
+}
+
+/**
+ * Connect all route handlers found to their respective routes
+ */
+export function connectRouteHandlers(app: Application, routes: RouteMap, context: RouteInitContext): Application {
+  for (const route in routes) {
+    const routePath = path.join('/api', route)
+
+    const { GET, POST, PUT, PATCH, DELETE, WEB_SOCKET } = routes[route].handlers
+
+    if (GET) {
+      connectRouteHandler(app, routePath, GET, context)
+    }
+
+    if (POST) {
+      connectRouteHandler(app, routePath, POST, context)
+    }
+
+    if (PUT) {
+      connectRouteHandler(app, routePath, PUT, context)
+    }
+
+    if (PATCH) {
+      connectRouteHandler(app, routePath, PATCH, context)
+    }
+
+    if (DELETE) {
+      connectRouteHandler(app, routePath, DELETE, context)
+    }
+
+    if (WEB_SOCKET) {
+      connectWebsocketHandler(app, routePath, WEB_SOCKET, context)
+    }
+  }
+
+  return app
+}
+
+/*
+ * Connect an exported route handler like GET, POST, etc to a specific express path.
+ */
+function connectRouteHandler(app: Application, routePath: string, func: RouteHandler, initContext: RouteInitContext): Application {
+  const handlerFunc = (req, res): void => {
```

Comment:
Could `req, res` be type as `res: Request, res: Response` instead of inferred from usage?

Reply:
Good catch!